File: Algorithms/p367_Valid_Perfect_Square_2.cpp
class Solution {
public:
    bool isPerfectSquare(int num) {
        for(int l=1,r=num+1;l<r;){
            int mid = (l+r)/2;
            int q = num/mid;
            if(mid<q)      l = mid+1;
            else if(mid>q) r = mid;
            else           return mid*mid==num;
        }
        return 0;
    }
};

File: Algorithms/p367_Valid_Perfect_Square_2.py
class Solution(object):
    def isPerfectSquare(self, num):
        """
        :type num: int
        :rtype: bool
        """
        l = 1
        r = num+1
        while l<r:
            mid = (l+r)//2
            q = num//mid
            if mid<q:
                l = mid+1
            elif mid>q:
                r = mid
            else:
                return mid**2==num
        return False
